tests/pass/libc.rs

```diff
@@ -311,11 +311,50 @@ fn test_posix_gettimeofday() {
     assert_eq!(is_error, -1);
 }
 
+fn test_posix_mkstemp() {
+    use std::ffi::CString;
+    use std::ffi::OsStr;
+    use std::fs::File;
+    use std::os::unix::ffi::OsStrExt;
+    use std::os::unix::io::FromRawFd;
+    use std::path::Path;
+
+    let valid_template = "fooXXXXXX";
+    // C needs to own this as `mkstemp(3)` says:
+    // "Since it will be modified, `template` must not be a string constant, but
+    // should be declared as a character array."
```

Comment:
Might be worth adding a `tests/fail` test to ensure we report UB when this *does* point to a string constant?

Reply:
```rust
let s: *mut libc::c_char = "hello".as_bytes() as *const _ as *mut _;
```

Reply:
Ah you need a trailing 0 as well I guess?
```rust
let s: *mut libc::c_char = b"fooXXXXXX\0" as *const _ as *mut _;
```

Reply:
When I try to get the type, it only sees the `*mut`. How can I get provenance / follow the pointer chain? The following doesn't throw:

```rust
        match template_op.layout.ty.kind() {
            ty::RawPtr(tym) =>
                if tym.mutbl != Mutability::Mut {
                    let (alloc_id, _, _) = this.ptr_get_alloc_id(ptr)?;
                    throw_ub!(WriteToReadOnly(alloc_id));
                },
            _ => unreachable!(),
        };
```

I see types like `Permission` but I can't seem to figure out how to get to it.

Reply:
You shouldn't do any of that in your shim. We already have the required checks elsewhere in Miri.

Reply:
Help me out? I don't see any mutablity checks on shim arg pointers and putting in a dumb test doesn't fail with miri's standard checks:


```rust
//@ignore-windows: No libc on Windows
//@compile-flags: -Zmiri-disable-isolation

#![feature(rustc_private)]

extern crate libc;

fn main() {
    test_mkstemp_immutable_arg();
}

fn test_mkstemp_immutable_arg() {
    let s: *mut libc::c_char = b"fooXXXXXX\0" as *const _ as *mut _;
    let _fd = unsafe { libc::mkstemp(s) };
}
```

gives me:

```bash
Fail got exit status: 0
```

So I need to add another assertion somewhere?

Reply:
It should error for the same reason that this errors:
```rust
fn main() {
    let b = b"fooXXXXXX\0";
    unsafe { (b.as_ptr() as *mut u8).write(0); }
}
```
Doesn't the implementation write to the pattern string?

Reply:
> The last six characters of template must be "XXXXXX" and these are replaced with a string that makes the filename unique.

is what the docs say. Have you implemented that part?

Reply:
I wasn't writing back to the pointer, doh. Fixed.